archunit/src/main/java/com/tngtech/archunit/core/importer/DomainBuilders.java

```diff
@@ -639,34 +657,74 @@ String getName() {
-            for (Map.Entry<JavaTypeVariable<JavaClass>, JavaTypeParameterBuilder<JavaClass>> typeParameterToBuilder : typeArgumentsToBuilders.entrySet()) {
+            Set<JavaTypeVariable<?>> allGenericParametersInContext = union(typeParametersFromEnclosingContextOf(owner), typeArgumentsToBuilders.keySet());
+            for (Map.Entry<JavaTypeVariable<OWNER>, JavaTypeParameterBuilder<OWNER>> typeParameterToBuilder : typeArgumentsToBuilders.entrySet()) {
                 List<JavaType> upperBounds = typeParameterToBuilder.getValue().getUpperBounds(allGenericParametersInContext);
                 completeTypeVariable(typeParameterToBuilder.getKey(), upperBounds);
             }
             return ImmutableList.copyOf(typeArgumentsToBuilders.keySet());
         }
+
+        abstract Set<JavaTypeVariable<?>> typeParametersFromEnclosingContextOf(OWNER owner);
     }
 
-    private static Set<JavaTypeVariable<?>> allTypeParametersInEnclosingClassesOf(JavaClass javaClass) {
+    static class JavaClassTypeParametersBuilder extends AbstractTypeParametersBuilder<JavaClass> {
+        JavaClassTypeParametersBuilder(List<JavaTypeParameterBuilder<JavaClass>> typeParameterBuilders) {
+            super(typeParameterBuilders);
+        }
+
+        @Override
+        Set<JavaTypeVariable<?>> typeParametersFromEnclosingContextOf(JavaClass javaClass) {
+            return allTypeParametersInEnclosingContextOf(javaClass);
+        }
+    }
+
+    static class JavaCodeUnitTypeParametersBuilder extends AbstractTypeParametersBuilder<JavaCodeUnit> {
+        JavaCodeUnitTypeParametersBuilder(List<JavaTypeParameterBuilder<JavaCodeUnit>> typeParameterBuilders) {
+            super(typeParameterBuilders);
+        }
+
+        @Override
+        Set<JavaTypeVariable<?>> typeParametersFromEnclosingContextOf(JavaCodeUnit codeUnit) {
+            return allTypeParametersInEnclosingContextOf(codeUnit);
+        }
+    }
+
+    private static Set<JavaTypeVariable<?>> allTypeParametersInEnclosingContextOf(JavaCodeUnit codeUnit) {
+        JavaClass declaringClass = codeUnit.getOwner();
+        return FluentIterable.from(getTypeParametersOf(declaringClass))
+                .append(allTypeParametersInEnclosingContextOf(declaringClass))
+                .toSet();
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static List<JavaTypeVariable<?>> getTypeParametersOf(HasTypeParameters<?> hasTypeParameters) {
+        List<? extends JavaTypeVariable<?>> result = hasTypeParameters.getTypeParameters();
+        return (List) result;
+    }
+
+    private static Set<JavaTypeVariable<?>> allTypeParametersInEnclosingContextOf(JavaClass javaClass) {
         Set<JavaTypeVariable<?>> result = new HashSet<>();
         while (javaClass.getEnclosingClass().isPresent()) {
+            if (javaClass.getEnclosingCodeUnit().isPresent()) {
```

Comment:
But for me that is no implementation detail, but domain intrinsic :thinking: It's just not possible by the JLS to have an enclosing method / constructor, but no enclosing class. Because every code unit that could be the enclosing declaration must be declared within a class which then by definition is an enclosing class, no? (I think that's also how the Reflection API treats it, I doubt you can construct an example where the Reflection API would give you an enclosing method, but no enclosing class)

Reply:
I mean, I can leave a comment there explaining that maybe, since it might be confusing in the future :+1: